### ISSUE TITLE
Feature Request: Extract data from job to use as job Title

### ISSUE THREAD

**Issue by bugs181** [felixmosh/bull-board#349]
A few versions back, I was "abusing" the named job parameter to have better titles on the dashboard. After upgrading to the latest; when named jobs are used we get the lovely "Missing process handler" error. I'll try working on a PR to add support to make this a little more generic which can be useful for other cases.

Original named jobs method that I was abusing to accomplish this goal:
```js
await bullQueue.add(jobItem.title, jobItem)
```

---

### Proposal 1:

Append `title` parameter to `QueueOptions`, which is just a generic JSON extractor.

Example:
```js
const bullQueue = new Bull('job-queue', { title: "jobTitle" })

...

const jobItem = {
  jobTitle: item.page.title,
  url: item.page.href,
}

await bullQueue.add(jobItem)
```

---

### Proposal 2:
Append `jobTitle?` parameter to `JobOptions`, which is just a generic JSON extractor.

Example:
```js
const bullQueue = new Bull('job-queue')

...

const jobItem = {
  jobTitle: item.page.title,
  url: item.page.href,
}

await bullQueue.add(jobItem, { jobTitle: "jobTitle" })
```

---

### Expected Output:

Expected title output would be something along the lines of:
```js
`${jobName} - ${jobTitle}`
```

---

### Bonus options:

- `replace?: boolean | undefined;` - Replace entire job title for the expected output section, instead of appending `jobTitle` to `jobName`.

**Comment by felixmosh** [comment]
Did you tried `options.prefix` ? I think that it's exactly what you need :]

**Comment by bugs181** [comment]
> Did you tried `options.prefix` ? I think that it's exactly what you need :]

Perhaps my wording wasn't clear or I am misunderstanding the use of prefix. I need the job titles to be unique per job. I was under the impression, the prefix is just static and used for all jobs. Am I incorrect?

**Comment by felixmosh** [comment]
Yes, you are correct.

Can you explain what do you want?

**Comment by bugs181** [comment]
The easiest explanation would be something along the lines of this:

```js
await bullQueue.add(jobItem, { jobTitle: "Downloading 123" })
await bullQueue.add(jobItem, { jobTitle: "Unzipping 123" })
await bullQueue.add(jobItem, { jobTitle: "Installing 123" })
```

I understand that example may still be confusing due to the above example, should have different queues - BUT it's just an example. I have a case where all jobs are the same type, but I want the titles to be unique at the time of adding them.

**Comment by felixmosh** [comment]
@bugs181 , I think that you've got confused, `bull-board` is a dashboard for your queues, It currently doesn't allows to add jobs to the queue, it only reflects it's data.

If you are using `bull` you can create a [job with a name](https://optimalbits.github.io/bull/#named-jobs) (which will be reflected on the dashboard).

if you are using `bullmq` you can create a job with a name like so: https://docs.bullmq.io/guide/jobs/job-ids.

**Comment by bugs181** [comment]
> @bugs181 , I think that you've got confused, `bull-board` is a dashboard for your queues, It currently doesn't allows to add jobs to the queue, it only reflects it's data.
> 
> If you are using `bull` you can create a [job with a name](https://optimalbits.github.io/bull/#named-jobs) (which will be reflected on the dashboard).
> 
> if you are using `bullmq` you can create a job with a name like so: https://docs.bullmq.io/guide/jobs/job-ids.

Unfortunately, I think there's a language barrier here. I'm fully aware of what bull-dashboard is. Displaying job data is exactly the purpose of the bull-board. Named jobs do not fulfill this purpose. I believe that a more generic approach, allowing users to dive into the job data to retrieve specific data for their job title on bull dashboard is important.

Pictures are worth a thousand words, so let me show you what I'm trying to accomplish instead.
![bull-board](https://user-images.githubusercontent.com/3699516/145148802-7b485af7-03de-40b9-8943-77fe7262cec3.png)

**Comment by felixmosh** [comment]
Ok, I think that i get it, what is your use case for this feature?

**Comment by bugs181** [comment]
Many more use cases than the one example above. Running a custom `getTitle` function would make this feature request more generic so it fits all my use cases and then some.

By using an `option` we can detect this `option` in `queues->formatJob` and override the `name` property.

A proposal example could be useful here:

```js
const queueOptions = {
  getTitle: function(jobProps) {
    return jobProps.data.title ? jobProps.data.title : jobProps.name
  }
}

createBullBoard({
  queues: [new BullAdapter(bullQueue, queueOptions)],
  serverAdapter,
})
```

Example titles could be different job data exposed for staged environments, job paths that are taken to arrive at the queue, etc. Basically, any kind of data that could be useful at a glance.

One such use-case I have is exposing Drone CI repo names.

This would make my titles look like these (in different queues):
- `Cloning bugs181/ML-Target`
- `Building bugs181/ML-Target`
- `Deploying bugs181/ML-Target`

**Comment by felixmosh** [comment]
We do have a built in formatters pattern that applies only on `data` & `returnValue` job props.

If I will add a support for formatting the `name` prop, will it help your use-case?

So, you will able to pass a formatter for the `name` field like so:
```ts
// server.js

const Queue = require('bull')
const QueueMQ = require('bullmq')
const { setQueues, BullMQAdapter, BullAdapter } = require('bull-board')

const someQueue = new Queue()
const someOtherQueue = new Queue()
const queueMQ = new QueueMQ()

const someQueueAdapter = new BullAdapter(someQueue);

// This is the proposed api
someQueueAdapter.setFormatter('name', function(jobProps) {
    const repo = jobProps.data
    return repo.action && repo.name ? `${repo.action} repo ${repo.name}` : jobProps.name
  }); // this will apply name formatter

setQueues([
  someQueueAdapter,
  new BullAdapter(someOtherQueue),
  new BullMQAdapter(queueMQ),
]);
```

WDYT?

**Comment by bugs181** [comment]
Your proposed solution could work. Not sure what other implications there are of formatting the data, though. It might be tricky to evaluate where the data came from if it modifies the `data` object in the dashboard, however. With my proposed solution, and PR #358 - we know exactly how the formatting happens and that it's only purpose is for display.

**Comment by felixmosh** [comment]
I don't agree, we are already have the concept of formatting other fields on the `job` object, `name` is one of them.
And it is better to follow the same convention instead of creating a new one.

**Comment by bugs181** [comment]
I think I see one more issue with `someQueueAdapter`. Will the formatter be applied to all queues or just the one? The reasoning this couldn't work for my use-case is when I have a dashboard where titles are very specific to the queue flow.

**Comment by felixmosh** [comment]
1. Formatters are queue based
2. Formatters are existing pattern
3. It is a matter of code style, you don't like mine and I don't think that yours is much better :]

If it solves your issue, I will make a PR, if you want to help, It will be faster, but I do want to continue the same existing pattern.

**Comment by bugs181** [comment]
That's acceptable. Code style aside, to each their own. No complaints here either way as long as we can accomplish the same task. I had a look at `setFormatter` but couldn't see how this would be applied to the `name` property. I'm largely unfamiliar with the code base right now. I'll keep taking a swing at it; but I have doubts I'll be able to make it work with my limited knowledge of the code.

**Comment by felixmosh** [comment]
No problem, I will make a PR. :]

**Comment by bugs181** [comment]
Doh! Silly me. I was over-complicating it.

This should do the trick:
`name: queue.format('name', jobProps),`

Unless you see something I missed, I'll revert the PR and apply this patch.

**Comment by felixmosh** [comment]
Yeap,

> Doh! Silly me. I was over-complicating it.
> 
> This should do the trick: `name: queue.format('name', jobProps),`
> 
> Unless you see something I missed, I'll revert the PR and apply this patch.

Yeap :]

**Comment by bugs181** [comment]
Thanks so much for your help! This looks great!

**Comment by felixmosh** [comment]
Thank you for you patience

**Comment by felixmosh** [comment]
Released in `v3.8.0`
